pkg/networkservice/chains/client/options.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/null"
+)
+
+type clientOptions struct {
+	name                    string
+	clientURL               *url.URL
+	cc                      grpc.ClientConnInterface
+	additionalFunctionality []networkservice.NetworkServiceClient
+	authorizeClient         networkservice.NetworkServiceClient
+	refreshClient           networkservice.NetworkServiceClient
+	dialOptions             []grpc.DialOption
+	dialTimeout             time.Duration
+}
+
+// Option modifies default client chain values.
+type Option func(c *clientOptions)
+
+// WithName sets name for the client.
+func WithName(name string) Option {
+	return Option(func(c *clientOptions) {
+		c.name = name
+	})
+}
+
+// WithClientURL sets name for the client.
+func WithClientURL(clientURL *url.URL) Option {
+	return Option(func(c *clientOptions) {
+		c.clientURL = clientURL
+	})
+}
+
+// WithClientConn sets name for the client.
+func WithClientConn(cc grpc.ClientConnInterface) Option {
+	return Option(func(c *clientOptions) {
+		c.cc = cc
+	})
+}
+
+// WithAdditionalFunctionality sets additionalFunctionality for the client. Note: this adds into tail of the client chain.
+func WithAdditionalFunctionality(additionalFunctionality ...networkservice.NetworkServiceClient) Option {
+	return Option(func(c *clientOptions) {
+		c.additionalFunctionality = additionalFunctionality
+	})
+}
+
+// WithAuthorizeClient sets authorizeClient for the client chain.
+func WithAuthorizeClient(authorizeClient networkservice.NetworkServiceClient) Option {
+	if authorizeClient == nil {
+		panic("authorizeClient cannot be nil")
+	}
+	return Option(func(c *clientOptions) {
+		c.authorizeClient = authorizeClient
+	})
+}
+
+// WithDialOptions sets dial options
+func WithDialOptions(dialOptions ...grpc.DialOption) Option {
+	return Option(func(c *clientOptions) {
+		c.dialOptions = dialOptions
+	})
+}
+
+// WithDialTimeout sets dial timeout
+func WithDialTimeout(dialTimeout time.Duration) Option {
+	return func(c *clientOptions) {
+		c.dialTimeout = dialTimeout
+	}
+}
+
+// WithoutRefresh disables refresh
+func WithoutRefresh() Option {
```

Comment:
What if we have different token timeout for the different applications?
```
NSC(1h) -> NSMgr(30m) -> Forwarder(15m) -> NSMgr(30m) -> NSE
```
If we use such option on `NSMgr, Forwarder` we should probably compute the least token timeout in `refresh` on `NSC` to become sure that there will be no timeouts in the chain.

Reply:
Probably it can be some rare case for the local scenario, because most probably all of the applications share the same spire server with same SVID TTL for all of them, but it also can be an interdomain scenario with different spire federations and so different SVID TTLs for local and remote applications.

Reply:
@Bolodya1997 You are completely correct we need to watch out for this.  We could do as you suggested and compute that in refresh.  I'm tempted though to do it in updatetoken by clamping down expiretime to the smallest in the path on the return direction of the path.